src/neo/SmartContract/ApplicationEngine.Runtime.cs

```diff
@@ -104,7 +106,20 @@ internal bool CheckWitnessInternal(UInt160 hash)
         {
             if (ScriptContainer is Transaction tx)
             {
-                if (!tx.Cosigners.TryGetValue(hash, out Cosigner cosigner)) return false;
+                IReadOnlyDictionary<UInt160, Cosigner> cosigners;
+                OracleResponse response = tx.Attributes.OfType<OracleResponse>().FirstOrDefault();
+                if (response is null)
+                {
+                    cosigners = tx.Cosigners;
+                }
+                else
+                {
+                    OracleRequest request = NativeContract.Oracle.GetRequest(Snapshot, response.Id);
+                    Transaction request_tx = Snapshot.GetTransaction(request.Txid);
+                    if (request_tx == null) return false;
+                    cosigners = request_tx.Cosigners;
```

Comment:
Union with the response? why they are required?

Reply:
The response doesn't need any cosigner. Because the respsone script is fixed.

https://github.com/neo-project/neo/blob/6d58ce91e3a33446193f7f96adb69ad3b325fec6/src/neo/Network/P2P/Payloads/Transaction.cs#L218-L219

But the callback needs cosigners. And it only needs the cosigners from the request. You can think of callback as continuing execution after the request is interrupted. So the callback cosigners should be the same as the request.